### ISSUE TITLE
P2P Network and Directory

### ISSUE THREAD

**Issue by matsjj** [matsjj/thundernetwork#5]
Due to the unresolved issue of efficient routing, and also because I think that such a service would emerge naturally on its own, I thought about a Node Directory, a central database where nodes can publish various data. It is completely optional, but I think it will help for routing and figuring out, to which nodes new nodes / clients should connect to. 

Data collected: 
- current peers the node is connected to
- fees for a payment towards another node
- funds available for payments (up to the node, how much information it gives)
- web-of-trust for node-rating. (uptime, attacks?, ping, ...)
- general information about new nodes (general location, ...)

Usage:
- open for public, clients and nodes can choose nodes based on ratings
- a central repository for efficient routing
- peer discovery for fresh nodes/clients
- nodes can choose to serve a connection if current fees are very high
- clients can gain from arbitrage, for routes with negative fees

Redundancy:
- we don't want this to be a single point of failure
- have multiple NDs, another P2P network between the NDs maybe

Realization:
- basically a big database
- every entry is signed with the pubkey from the node
- entries can only be changed by the nodes who entered them
- similar to the blockchain? 

![drawing1](https://cloud.githubusercontent.com/assets/4670837/9662669/b776b730-5260-11e5-9a89-075e9c542400.jpg)

**Comment by matsjj** [comment]
Was a while since I initially proposed this. I guess given the little amount of data that needs to be exchanged, we are good for now without a central directory, but rather share all the data between the nodes. In case this ever becomes a problem, nodes with bandwidth constraints can always make use of the - by then - established central directories.
